File: frontend/src/lib/rules.js
// Detection engine ported from prototype.html (findConflicts / getIssues).
// PRD 10 names this the canonical behaviour. Once GET /patients returns `issues`,
// prefer the server engine so cards and brief cannot drift.
import { TOPICS, FIELDS, TYPE_LABEL, W } from '../data/chart.js'

// Hours read fine up to a day; past that "44h" makes you do arithmetic to see
// it is not today. Switch to days + hours at the 24h boundary.
export function etaLabel(h) {
  if (h < 24) return `${h}h`
  const d = Math.floor(h / 24)
  const rem = h % 24
  return rem ? `${d}d ${rem}h` : `${d}d`
}

export function allNotes(p, state) {
  return p.notes.concat(state.added[p.id] || [])
}

export function tagOf(n, topic) {
  const t = n.tags.find((x) => x[0] === topic)
  return t ? t[1] : null
}

// Deterministic coordination suggestion for offline/mock mode — mirrors the
// fallback branch of backend/app/suggest.py. Process steps only, never a
// clinical answer, and never a winner when instructions conflict.
export function suggestFor(issue) {
  let text
  if (issue.type === 'conflict') {
    const roles = [...new Set(
      Object.values(issue.vals || {}).flat().map((n) => n.role).filter(Boolean),
    )].sort()
    text = `Have the attending reconcile this one — ask ${roles.join(' and ') || 'each team'} ` +
      'to confirm which instruction was last carried out before anyone follows ' +
      'either. Baton flags conflicts; it does not pick the winner.'
  } else if (issue.type === 'handoff') {
    text = issue.fix?.kind === 'pending'
      ? 'Assign a named owner for this pending result before shift change — ' +
        'unowned follow-ups are the ones that come back abnormal and go unseen.'
      : 'Fill this field before the next shift takes over — it is required ' +
        'for a complete handoff.'
  } else {
    text = issue.escalated
      ? 'Escalation is logged — confirm the charge nurse or bed coordinator ' +
        'has seen it, and note the expected resolution time in the handoff.'
      : 'Assign an owner and confirm with whoever it is waiting on; if it has ' +
        'not moved by the next check-in, escalate to the charge nurse.'
  }
  return { text, source: 'rules' }
}

export function ago(n) {
  if (n.at) {
    const m = Math.round((Date.now() - n.at) / 60000)
    return m < 1 ? 'just now' : m < 60 ? `${m}m ago` : `${Math.round(m / 60)}h ago`
  }
  return `${n.h}h ago`
}

export function findConflicts(p, state) {
  const notes = allNotes(p, state)
  const out = []
  Object.keys(TOPICS).forEach((topic) => {
    const tn = notes.filter((n) => tagOf(n, topic) !== null)
    if (!tn.length) return
    // Adopting a value writes a reconciling note; older instructions stop counting.
    let barrier = 0
    tn.forEach((n) => { if (n.reconcile && n.seq > barrier) barrier = n.seq })
    const latest = {}
    tn.filter((n) => n.seq >= barrier).forEach((n) => {
      if (!latest[n.role] || n.seq > latest[n.role].seq) latest[n.role] = n
    })
    const vals = {}
    Object.keys(latest).forEach((r) => {
      const n = latest[r]
      const v = tagOf(n, topic)
      ;(vals[v] = vals[v] || []).push(n)
    })
    if (Object.keys(vals).length > 1) out.push({ topic, vals })
  })
  return out
}

export function getIssues(p, state) {
  const issues = []

  findConflicts(p, state).forEach((c) => {
    const T = TOPICS[c.topic]
    const vs = Object.keys(c.vals)
    const sub = []
    vs.forEach((v) => c.vals[v].forEach((n) => sub.push(`${n.role}: ${v}`)))
    issues.push({
      id: `${p.id}:conflict:${c.topic}`, pid: p.id, type: 'conflict', sev: T.sev,
      title: `${T.label}: ${vs.length} different instructions are active`,
      sub: sub.join('; '),
      why: 'Both cannot be followed. Until someone picks one, whoever reads the chart next may follow either.',
      topic: c.topic, vals: c.vals,
    })
  })

  FIELDS.forEach((f) => {
    let v = (state.filled[p.id] || {})[f.key]
    if (v === undefined) v = p.handoff[f.key]
    if (!v) {
      issues.push({
        id: `${p.id}:handoff:${f.key}`, pid: p.id, type: 'handoff', sev: f.sev,
        title: `${f.label} is missing from the handoff`,
        sub: f.why, why: f.why, fix: { kind: 'field', key: f.key, ph: f.ph },
      })
    }
  })

  p.pending.forEach((r) => {
    const o = state.pendOwners[`${p.id}|${r.name}`] || r.owner
    if (!o) {
      issues.push({
        id: `${p.id}:handoff:pend:${r.name}`, pid: p.id, type: 'handoff', sev: 'med',
        title: `Pending result has no owner: ${r.name}`,
        sub: 'Result pending, nobody assigned to follow it up',
        why: 'If nobody owns a pending result, it can come back abnormal and go unseen after the patient leaves.',
        fix: { kind: 'pending', name: r.name },
      })
    }
  })

  p.blockers.forEach((b) => {
    if ((state.cleared[p.id] || {})[b.id]) return
    let sev = 'med'
    if (b.blocks && (p.dischargeInH <= 24 || b.ageH >= 24)) sev = 'high'
    else if (!b.blocks && b.ageH < 24) sev = 'low'
    issues.push({
      id: `${p.id}:blocker:${b.id}`, pid: p.id, type: 'blocker', sev,
      title: b.label,
      sub: `Waiting on ${b.waitingOn} for ${b.ageH}h. Discharge target in ${etaLabel(p.dischargeInH)}.`,
      why: b.cat + (b.blocks ? ' item that blocks discharge.' : ' item.'),
      bid: b.id, escalated: !!state.escalated[`${p.id}|${b.id}`],
    })
  })

  issues.sort((a, b) => W[b.sev] - W[a.sev])
  return issues
}

export function riskOf(issues) {
  return issues.reduce((a, i) => a + W[i.sev], 0)
}

export function dischStatus(p, issues) {
  if (issues.some((i) => i.sev === 'high')) return 'risk'
  return issues.length ? 'watch' : 'ready'
}

// Takes an already-overlaid patient, so it works the same on API and mock data.
export function handoffProgress(p) {
  const total = FIELDS.length + p.pending.length
  let done = 0
  FIELDS.forEach((f) => { if (p.handoff[f.key]) done++ })
  p.pending.forEach((r) => { if (r.owner) done++ })
  return { done, total, pct: Math.round((done / total) * 100) }
}

// Unit-wide brief. PRD 8.1 pins GET /brief to this shape: severity, then soonest
// discharge, cap 14 with a remainder line. Never written back to the chart.
export function briefText(patients, state) {
  const all = []
  patients.forEach((p) => getIssues(p, state).forEach((i) => all.push({ p, i })))
  all.sort((a, b) => W[b.i.sev] - W[a.i.sev] || a.p.dischargeInH - b.p.dischargeInH)
  const un = all.filter((x) => !state.owners[x.i.id]).length
  const L = [
    'SHIFT HANDOFF BRIEF (synthetic data)',
    `${all.length} open coordination issues, ${un} with no owner.`,
    '',
  ]
  if (!all.length) L.push('Nothing open. Instructions agree, handoffs are complete, no blockers.')
  all.slice(0, 14).forEach((x, k) => {
    L.push(`${k + 1}. [${x.i.sev.toUpperCase()}] Rm ${x.p.room} ${x.p.name}: ${TYPE_LABEL[x.i.type]}. ${x.i.title}.`)
    L.push(`   ${x.i.sub}`)
    L.push(`   Owner: ${state.owners[x.i.id] || 'UNASSIGNED'}; discharge target in ${etaLabel(x.p.dischargeInH)}.`)
    L.push('')
  })
  if (all.length > 14) L.push(`+ ${all.length - 14} lower-priority items in Baton.`)
  return L.join('\n')
}

File: backend/app/rules.py
"""Rule engine ported from frontend/public/prototype.html (lines 269-429).

Data model: patient dict with keys id, name, dischargeInH, handoff{...},
pending[{name, owner}], blockers[{id, label, cat, waitingOn, ageH, blocks}],
notes[{role, author, h, text, tags: [[topic, value], ...], reconcile?, seq?}].

`state` mirrors the prototype's persisted dicts:
    filled:      {pid: {fieldKey: value}}
    pendOwners:  {"pid|result name": owner}
    cleared:     {pid: {blockerId: true}}
    escalated:   {"pid|blockerId": true}
    added:       {pid: [notes appended after load]}
"""

TOPICS = {
    "anticoagulation": {"label": "Anticoagulation", "values": ["continue", "hold"], "sev": "high"},
    "weight_bearing": {"label": "Weight-bearing", "values": ["full", "partial", "non-weight-bearing"], "sev": "high"},
    "destination": {"label": "Discharge destination", "values": ["home", "home with home health", "SNF", "inpatient rehab"], "sev": "high"},
    "diet": {"label": "Diet", "values": ["regular", "cardiac 2g Na", "carb-consistent", "NPO", "clear liquids"], "sev": "med"},
    "fluids": {"label": "Fluids", "values": ["restrict", "encourage"], "sev": "med"},
    "isolation": {"label": "Isolation", "values": ["contact precautions", "none"], "sev": "med"},
}

FIELDS = [
    {"key": "codeStatus", "label": "Code status", "sev": "high",
     "why": "Oncoming staff cannot act safely in an emergency without it.",
     "ph": "e.g. Full code"},
    {"key": "allergies", "label": "Allergies", "sev": "high",
     "why": "Allergy status is not documented in the handoff, so orders may be placed without it.",
     "ph": "e.g. NKDA, or drug and reaction"},
    {"key": "followUpOwner", "label": "Post-discharge follow-up owner", "sev": "med",
     "why": "Nobody is named to see this patient after discharge, so follow-up is likely to slip.",
     "ph": "e.g. PCP Dr. Hale"},
    {"key": "medRec", "label": "Medication reconciliation", "sev": "med",
     "why": "Discharge medications have not been reconciled against home medications.",
     "ph": "e.g. Done by pharmacy"},
    {"key": "familyContact", "label": "Family or surrogate contact", "sev": "low",
     "why": "No one is listed to call if the patient's status changes or a plan needs consent.",
     "ph": "e.g. Daughter, on file"},
]

W = {"high": 3, "med": 2, "low": 1}


def empty_state() -> dict:
    return {"added": {}, "filled": {}, "pendOwners": {}, "cleared": {}, "escalated": {}}


def assign_seq(notes: list[dict]) -> list[dict]:
    """Sort notes newest-first (h desc) and assign seq = 1..n, as the prototype does."""
    notes.sort(key=lambda n: -n["h"])
    for i, n in enumerate(notes):
        n["seq"] = i + 1
    return notes


def all_notes(p: dict, state: dict) -> list[dict]:
    return list(p["notes"]) + state.get("added", {}).get(p["id"], [])


def tag_of(note: dict, topic: str):
    for t in note.get("tags", []):
        if t[0] == topic:
            return t[1]
    return None


def find_conflicts(p: dict, state: dict | None = None) -> list[dict]:
    state = state or empty_state()
    notes = all_notes(p, state)
    out = []
    for topic in TOPICS:
        tn = [n for n in notes if tag_of(n, topic) is not None]
        if not tn:
            continue
        barrier = max((n["seq"] for n in tn if n.get("reconcile")), default=0)
        latest: dict[str, dict] = {}
        for n in tn:
            if n["seq"] >= barrier and (n["role"] not in latest or n["seq"] > latest[n["role"]]["seq"]):
                latest[n["role"]] = n
        vals: dict[str, list[dict]] = {}
        for n in latest.values():
            vals.setdefault(tag_of(n, topic), []).append(n)
        if len(vals) > 1:
            out.append({"topic": topic, "vals": vals})
    return out


def get_issues(p: dict, state: dict | None = None) -> list[dict]:
    S = state or empty_state()
    issues = []
    for c in find_conflicts(p, S):
        T = TOPICS[c["topic"]]
        vs = list(c["vals"].keys())
        sub = [f"{n['role']}: {v}" for v in vs for n in c["vals"][v]]
        issues.append({
            "id": f"{p['id']}:conflict:{c['topic']}", "pid": p["id"], "type": "conflict",
            "sev": T["sev"],
            "title": f"{T['label']}: {len(vs)} different instructions are active",
            "sub": "; ".join(sub),
            "why": "Both cannot be followed. Until someone picks one, whoever reads the chart next may follow either.",
            "topic": c["topic"], "vals": c["vals"]})
    for f in FIELDS:
        v = S.get("filled", {}).get(p["id"], {}).get(f["key"])
        if v is None:
            v = p["handoff"].get(f["key"])
        if not v:
            issues.append({
                "id": f"{p['id']}:handoff:{f['key']}", "pid": p["id"], "type": "handoff",
                "sev": f["sev"],
                "title": f"{f['label']} is missing from the handoff",
                "sub": f["why"], "why": f["why"],
                "fix": {"kind": "field", "key": f["key"], "ph": f["ph"]}})
    for r in p.get("pending", []):
        o = S.get("pendOwners", {}).get(f"{p['id']}|{r['name']}") or r.get("owner")
        if not o:
            issues.append({
                "id": f"{p['id']}:handoff:pend:{r['name']}", "pid": p["id"], "type": "handoff",
                "sev": "med",
                "title": f"Pending result has no owner: {r['name']}",
                "sub": "Result pending, nobody assigned to follow it up",
                "why": "If nobody owns a pending result, it can come back abnormal and go unseen after the patient leaves.",
                "fix": {"kind": "pending", "name": r["name"]}})
    for b in p.get("blockers", []):
        if S.get("cleared", {}).get(p["id"], {}).get(b["id"]):
            continue
        sev = "med"
        if b["blocks"] and (p["dischargeInH"] <= 24 or b["ageH"] >= 24):
            sev = "high"
        elif not b["blocks"] and b["ageH"] < 24:
            sev = "low"
        issues.append({
            "id": f"{p['id']}:blocker:{b['id']}", "pid": p["id"], "type": "blocker",
            "sev": sev, "title": b["label"],
            "sub": f"Waiting on {b['waitingOn']} for {b['ageH']}h. Discharge target in {p['dischargeInH']}h.",
            "why": b["cat"] + (" item that blocks discharge." if b["blocks"] else " item."),
            "bid": b["id"],
            "escalated": bool(S.get("escalated", {}).get(f"{p['id']}|{b['id']}"))})
    issues.sort(key=lambda i: -W[i["sev"]])
    return issues


def risk_of(issues: list[dict]) -> int:
    return sum(W[i["sev"]] for i in issues)


def disch_status(p: dict, issues: list[dict]) -> str:
    if any(i["sev"] == "high" for i in issues):
        return "risk"
    return "watch" if issues else "ready"
